### PR TITLE
Makes select_speaker more robust by checking for mentions anywhere.

### LINKED ISSUE #663: GroupChat is overly sensitive to formatting when selecting the next speaker.
As pointed out by Jonny7377 on Discord, GroupChat is overly sensitive to formatting issues when selecting the next speaker. E.g., 

```
GroupChat select_speaker failed to resolve the next speaker's name. Speaker selection will default to the next speaker in the list. This is because the speaker selection OAI call returned:
 Coder
```

In this case, the extra space causes the lookup to fail. 

At the very least, the code should strip() whitespace from the agent names before doing a lookup. That can be done here:

https://github.com/microsoft/autogen/blob/4711260582536e23ebbd9c3ccd067cddf60cd21f/autogen/agentchat/groupchat.py#L42

Or here,

https://github.com/microsoft/autogen/blob/4711260582536e23ebbd9c3ccd067cddf60cd21f/autogen/agentchat/groupchat.py#L90

Alternatively, a more robust option might be to check if *exactly one* agent is mentioned in response, and if so, select it (regardless of how it is formatted).

### PR DESCRIPTION
**Pull request by afourney**
Makes select_speaker more robust by checking for agent name mentions anywhere in the string returned by the LLM.

## Why are these changes needed?

Selecting the speaker fails when the LLM returns anything other than an exact agent name.

## Related issue number

Closes #663, and perhaps a few others.

Edit: I just saw #668, which is a different problem, but might be related?

## Checks

- [ ] I've included any doc changes needed for https://microsoft.github.io/autogen/. See https://microsoft.github.io/autogen/docs/Contribute#documentation to build and test documentation locally.
- [ ] I've added tests (if relevant) corresponding to the changes introduced in this PR.
- [ ] I've made sure all auto checks have passed.

### CODE REVIEW

File: autogen/agentchat/groupchat.py [hunk 1]
```diff
@@ -100,6 +101,20 @@ def select_speaker(self, last_speaker: Agent, selector: ConversableAgent):
...
+        mentions = dict()
+        for agent in self.agents:
+            regex = r"\b" + re.escape(agent.name) + r"\b"  # Finds agent mentions, taking word boundaries into account
+            count = len(re.findall(regex, name))
+            if count > 0:
+                mentions[agent.name] = count
+
+        # If exactly one agent is found, use it. Otherwise, leave the OAI response unmodified
+        if len(mentions) == 1:
+            name = next(iter(mentions))
```
Comment (BeibinLi) [inline autogen/agentchat/groupchat.py:115]: How about using set?

```python
# Find mentioned agents, and set 'name' to the agent's name iff one agent is mentioned
mentioned = {agent.name for agent in self.agents if len(re.findall(r"\b" + re.escape(agent.name) + r"\b", name)) > 0}
name = next(iter(mentioned)) if len(mentioned) == 1 else name
```

**Review by BeibinLi — APPROVED** [review]
The change looks good to me.

I made a suggestion to use "set", but feel free to ignore it.

**Comment by afourney** [comment]
Thanks. I'm happy either way, but personally find complex list compressions take longer to read/understand. If it's all the same to you, I'd rather just leave it step by step.

File: autogen/agentchat/groupchat.py [hunk 2]
```diff
@@ -100,6 +101,20 @@ def select_speaker(self, last_speaker: Agent, selector: ConversableAgent):
...
             # i = self._random.randint(0, len(self._agent_names) - 1)  # randomly pick an id
             return self.next_agent(last_speaker, agents)
+
+        # Find mentions of any agents
+        mentions = dict()
+        for agent in self.agents:
+            regex = r"\b" + re.escape(agent.name) + r"\b"  # Finds agent mentions, taking word boundaries into account
+            count = len(re.findall(regex, name))
+            if count > 0:
+                mentions[agent.name] = count
```
Comment (sonichi) [inline autogen/agentchat/groupchat.py:111]:
```suggestion
                mentions[agent.name] = count
                if len(mentions) > 1:
                    break
```
early stop to be efficient

**Review by sonichi — COMMENTED** [review]
Could you add a test to cover the changed function?

File: autogen/agentchat/groupchat.py (hunk 2 above)
Comment (afourney) [reply autogen/agentchat/groupchat.py]: I refactored this logic to a separate function. This comment is no longer applicable.

**Comment by afourney** [comment]
> Could you add a test to cover the changed function?

Yes done. To test this effectively, I needed to separate some of the logic out as a helper function

**Review by qingyun-wu — APPROVED** [review]

File: autogen/agentchat/groupchat.py [hunk 3]
```diff
@@ -119,6 +126,22 @@ def _participant_roles(self):
...
+        Returns: A dictionary mapping agent names to mention counts (to be included, at least one mention must occur)
+        """
+        mentions = dict()
+        for agent in agents:
+            regex = (
+                r"(?<=\W)" + re.escape(agent.name) + r"(?=\W)"
+            )  # Finds agent mentions, taking word boundaries into account
+            count = len(re.findall(regex, " " + message_content + " "))  # Pad the message to help with matching
+            if count > 0:
+                mentions[agent.name] = count
```
Comment (sonichi) [inline autogen/agentchat/groupchat.py:142]: Early stop is still applicable here.

**Pull request merged**